app/lib/cluster_viz_service.rb

```diff
@@ -0,0 +1,252 @@
+        # Set color index that will be interpreted by SCP front end
+        cluster_props[:marker][:scpColorIndex] = index
+      end
+
+      plot_data.push(cluster_props)
+    end
+
+    plot_data
+  end
+
+    # load custom coordinate-based annotation labels for a given cluster
+  def self.load_cluster_group_coordinate_labels(cluster)
+    # assemble source data
+    x_array = cluster.concatenate_data_arrays('x', 'labels')
+    y_array = cluster.concatenate_data_arrays('y', 'labels')
+    z_array = cluster.concatenate_data_arrays('z', 'labels')
+    text_array = cluster.concatenate_data_arrays('text', 'labels')
+    annotations = []
+    # iterate through list of data objects to construct necessary annotations
+    x_array.each_with_index do |point, index|
+      annot = {
+        showarrow: false,
+        x: point,
+        y: y_array[index],
+
+        text: text_array[index],
+        font: {
+          family: cluster.coordinate_labels_options[:font_family],
+          size: cluster.coordinate_labels_options[:font_size],
+          color: cluster.coordinate_labels_options[:font_color]
+        }
+      }
+      if cluster.is_3d?
+        annot[:z] = z_array[index]
+      end
+      annotations << annot
+    end
+    annotations
+  end
+
+  # retrieve axis labels from cluster coordinates file (if provided)
+  def self.load_axis_labels(cluster)
+    coordinates_file = cluster.study_file
+    {
+        x: coordinates_file.x_axis_label.blank? ? 'X' : coordinates_file.x_axis_label,
+        y: coordinates_file.y_axis_label.blank? ? 'Y' : coordinates_file.y_axis_label,
+        z: coordinates_file.z_axis_label.blank? ? 'Z' : coordinates_file.z_axis_label
+    }
+  end
+
+  # compute the aspect ratio between all ranges and use to enforce equal-aspect ranges on 3d plots
+  def self.compute_aspect_ratios(range)
+    # determine largest range for computing aspect ratio
+    extent = {}
+    range.each.map {|axis, domain| extent[axis] = domain.first.upto(domain.last).size - 1}
+    largest_range = extent.values.max
+
+    # now compute aspect mode and ratios
+    aspect = {
+        mode: extent.values.uniq.size == 1 ? 'cube' : 'manual'
+    }
+    range.each_key do |axis|
+      aspect[axis.to_sym] = extent[axis].to_f / largest_range
+    end
+    aspect
+  end
+
+  # set the range for a plotly scatter, will default to data-defined if cluster hasn't defined its own ranges
+  # dynamically determines range based on inputs & available axes
+  def self.set_range(cluster, inputs)
+    # select coordinate axes from inputs
+    domain_keys = inputs.map(&:keys).flatten.uniq.select {|i| [:x, :y, :z].include?(i)}
+    range = Hash[domain_keys.zip]
+    if cluster.has_range?
+      # use study-provided range if available
+      range = cluster.domain_ranges
+    else
+      # take the minmax of each domain across all groups, then the global minmax
+      vals = inputs.map {|v| domain_keys.map {|k| RequestUtils.get_minmax(v[k])}}.flatten.minmax
```

Comment:
Since this has abstracted out of `site_controller.rb`, we need to get it up to par with what is on `development` now:

```
# take the minmax of each domain across all groups, then the global minmax
raw_values = []
inputs.each do |input|
  domain_keys.each do |domain|
    domain_range = RequestUtils.get_minmax(input[domain])
    # RequestUtils.get_minmax will discard NaN/nil values that were ingested
    # only add domain range to list if we have a valid minmax
    raw_values << domain_range if domain_range.any?
  end
end
aggregate_range = raw_values.flatten.minmax
# add 2% padding to range
padding = (aggregate_range.first - aggregate_range.last) * 0.02
absolute_range = [aggregate_range.first + padding, aggregate_range.last - padding]
range[:x] = absolute_range
range[:y] = absolute_range
range[:z] = absolute_range
```